cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/FixedWindowRollingPolicy.java

```diff
@@ -0,0 +1,151 @@
+  // TODO Support compression
+  private int minIndex;
+  private static final String FNP_NOT_SET =
+    "The \"FileNamePattern\" property must be set before using FixedWindowRollingPolicy. ";
+  private int maxIndex;
+
+  /**
+   * It's almost always a bad idea to have a large window size, say over 12.
+   */
+  private static final int MAX_WINDOW_SIZE = 12;
+
+  public FixedWindowRollingPolicy() {
+    minIndex = 1;
+    maxIndex = 7;
+  }
+
+  @Override
+  public void start() {
+    if (fileNamePatternStr != null) {
+      fileNamePattern = new FileNamePattern(fileNamePatternStr, this.context);
+    } else {
+      addError(FNP_NOT_SET);
+      addError(CoreConstants.SEE_FNP_NOT_SET);
+      throw new IllegalStateException(FNP_NOT_SET + CoreConstants.SEE_FNP_NOT_SET);
+    }
+
+    if (maxIndex < minIndex) {
+      addWarn("MaxIndex (" + maxIndex + ") cannot be smaller than MinIndex ("
+                + minIndex + ").");
+      addWarn("Setting maxIndex to equal minIndex.");
+      maxIndex = minIndex;
+    }
+
+    if ((maxIndex - minIndex) > MAX_WINDOW_SIZE) {
+      addWarn("Large window sizes are not allowed.");
+      maxIndex = minIndex + MAX_WINDOW_SIZE;
+      addWarn("MaxIndex reduced to " + maxIndex);
+    }
+
+    IntegerTokenConverter itc = fileNamePattern.getIntegerTokenConverter();
+
+    if (itc == null) {
+      throw new IllegalStateException("FileNamePattern ["
+                                        + fileNamePattern.getPattern()
+                                        + "] does not contain a valid IntegerToken");
+    }
+
+    super.start();
+  }
+
+  @Override
+  public void rollover() throws RolloverFailure {
+    Location parentLocation = Locations.getParent(activeFileLocation);
+
+    if (parentLocation == null) {
+      return;
+    }
+
+    // Inside this method it is guaranteed that the hereto active log file is
+    // closed.
+    // If maxIndex <= 0, then there is no file renaming to be done.
+    if (maxIndex >= 0) {
+      try {
+        // Delete the oldest file, to keep Windows happy.
+        String fileName = fileNamePattern.convertInt(maxIndex);
+        Location deleteLocation = parentLocation.append(fileName);
+        if (deleteLocation.exists()) {
+          deleteLocation.delete();
+        }
+
+        // Map {(maxIndex - 1), ..., minIndex} to {maxIndex, ..., minIndex+1}
+        for (int i = maxIndex - 1; i >= minIndex; i--) {
+          String toRenameStr = fileNamePattern.convertInt(i);
+          Location toRename = parentLocation.append(toRenameStr);
+          // no point in trying to rename an non existent file
+          if (toRename.exists()) {
+            Location newName = parentLocation.append(fileNamePattern.convertInt(i + 1));
+            toRename.renameTo(newName);
+          } else {
+            addInfo("Skipping roll-over for inexistent file " + toRenameStr);
+          }
+        }
+      } catch (IOException e) {
+        throw new RolloverFailure(e.getMessage());
```

Comment:
You can do 

```java
RolloverFailure f = new RolloverFailure(e.getMessage());
f.addSuppressed(e);
throw f;
```

to kind of preserve the information (and stacktrace) of the original failure.